Immersive Portals!

Jeryn99 opened the pull request:
This pull request was worked on by myself and @thejudge156 

It supplies Immersive Portals support for the mod for both the shell and the interior.
Shells that do not have any portal information given, will operate without portals like the mods default behaviour.

Fabric users simply need to install any version of Immersive Portals and Fabric API, but Forge users will need to use Immersive Portals 2.3.2 and onwards; using earlier versions will cause the game to close to avoid users pointing out a crash where adding a dimension via code would crash the game with a crash report.

The compatibility will be optional via the implemented config file (the only exception being on Fabric, where dimensions NEED to be handled via Immersive Portals) 


![2023-02-01_19 25 03](https://user-images.githubusercontent.com/36424503/216142860-095ec004-8a26-49b8-b10e-05020d55fe2c.png)
![2023-02-01_19 25 16](https://user-images.githubusercontent.com/36424503/216142863-cc0a74cc-d7cd-463e-8061-2856663608bc.png)
![2023-02-01_19 24 15](https://user-images.githubusercontent.com/36424503/216142864-5d90cc3a-4e86-42d2-a1f1-1aac796af0ab.png)
![2023-01-07_19 43 39](https://user-images.githubusercontent.com/36424503/216142888-11ea6ab5-4998-4420-b469-a398c3745930.png)

Jeryn99 committed:
IP stuff

thejudge156 committed:
Merge branch 'master' into feature/ip

# Conflicts:
#	fabric/src/main/java/whocraft/tardis_refined/fabric/events/ModEvents.java

thejudge156 committed:
Immersive Portals Work

Things that work:
Portals for all rotations of Exterior and Interior combinations,
Portals for all Exteriors and Interior doors, except Briefcase,
Smooth dimension change,
TARDIS Portal Data needs some rework.

Things that don't work:
Nothing to my knowledge, some stuff needs cleanup though.

thejudge156 committed:
Fix crash

thejudge156 committed:
Revert model changes :(

Jeryn99 committed:
god among men

thejudge156 committed:
Clean up & Improve Immersive Portals Intergration (#59)

Co-authored-by: Jacob <42728283+JacobKauffman@users.noreply.github.com>
Co-authored-by: Clayton T <34910888+CommandrMoose@users.noreply.github.com>
Co-authored-by: 50ap5ud5 <50ap5ud5125@gmail.com>
Co-authored-by: MagicMaan <38400400+Magicmaan@users.noreply.github.com>

Jeryn99 committed:
Made level with master

Jeryn99 committed:
Log compatibility triggering

Jeryn99 committed:
Merge

Jeryn99 committed:
Implemented config for IP

Jeryn99 committed:
Implemented config for IP, again, but this time not flipped

Jeryn99 committed:
Flag unregistered shells

thejudge156 committed:
Fix several issues with the current IP implementation (#70)

Jeryn99 committed:
Made level with master

Jeryn99 committed:
what

Jeryn99 committed:
Small refactor for work on Forge

Jeryn99 committed:
Server fixes

Jeryn99 committed:
Updated with origin

Jeryn99 committed:
TardisInternalDoor

Jeryn99 committed:
forge ip is not stable, forge ip is unhappy

Jeryn99 committed:
Merge remote-tracking branch 'origin/master' into feature/ip

Jeryn99 committed:
Minor changes

Jeryn99 committed:
Users will need to use 2.3.2 IP for Forge or the game will force close

Jeryn99 committed:
ShellThemes that do not have IP data, will act normally as if the mod is not installed

Jeryn99 committed:
Removed Control group listing

Jeryn99 committed:
You will obey me JitPack

Jeryn99 committed:
Updated textures to fix weird rendering issues

CommandrMoose reviewed (commented):
Feels way more solid than it has before. I have found a few bugs that are either in relation to or are more present when using this feature.

BUG: Leaving the game and rejoining whilst inside the cave makes the growth door no longer function.
BUG: Placing a door inside the root cave creates a portal for the growth shell
BUG: Default shell's portal is not aligned with the interior view - doors are distinctively lower inside when looking from the outside
BUG: You cannot view the external shell's model whilst inside - doors that open outwards are not visible.
BUG: Door sounds can be heard twice when the door is closed
BUG: Throwing items through the door will cause them to spaz out
BUG: Breaking the shell in creative causes the model to no longer render until the next interaction.

Before merging, please make sure that everything is operational and works as intended by building and using the TARDIS without immersive portals installed.

Suff99 commented on common/src/main/java/whocraft/tardis_refined/compat/portals/ImmersivePortals.java:
```diff
@@ -0,0 +1,259 @@
+            }
+        });
+
+        themeToOffsetMap.put(ShellTheme.FACTORY, List.of(new Vec3(0.499, 0, 0),
```
Done

Suff99 commented on common/src/main/java/whocraft/tardis_refined/compat/portals/TARDISPortalData.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.UUID;
+
+public interface TARDISPortalData {
```
Renamed

Suff99 commented on forge/src/main/java/whocraft/tardis_refined/common/capability/forge/TardisLevelOperatorImpl.java:
```diff
@@ -59,6 +59,12 @@ public static void onLevelTick(TickEvent.LevelTickEvent event) {
     public static Optional<TardisLevelOperator> get(ServerLevel level) {
+
+        if(level == null) {
+            System.out.println("what the fuck");
```
Removed

Suff99 commented on common/src/main/java/whocraft/tardis_refined/client/model/blockentity/shell/ShellModelCollection.java:
```diff
@@ -2,6 +2,7 @@
 
 import net.minecraft.client.Minecraft;
 import whocraft.tardis_refined.client.ModelRegistry;
+import whocraft.tardis_refined.client.model.blockentity.console.ConsoleModelCollection;
```
Removed

Suff99 commented on common/src/main/java/whocraft/tardis_refined/common/blockentity/door/GlobalDoorBlockEntity.java:
```diff
@@ -32,7 +33,7 @@ public void onRightClick(BlockState blockState, TardisInternalDoor door, Player
                     return;
                 }
-                if (!cap.getControlManager().isInFlight() && !door.locked()) {
+                if (!cap.getControlManager().isInFlight() && !door.locked() && serverLevel.dimensionTypeId().equals(DimensionTypes.TARDIS)) {
```
Removed

Suff99 commented on common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/VendingMachineDoorModel.java:
```diff
@@ -50,7 +50,7 @@ public ModelPart root() {
     @Override
     public void setDoorPosition(boolean open) {
-        this.door.xRot = (open) ? 1.6f : 0;
+        this.door.xRot = (open) ? -1.6f : 0;
```
Fixed

Suff99 commented on common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/PresentDoorModel.java:
```diff
@@ -66,7 +66,7 @@ public ModelPart root() {
 	@Override
 	public void setDoorPosition(boolean open) {
-		this.door.xRot = (open) ? 1.5f:0;
+		this.door.xRot = (open) ? -1.5f:0;
```
Fixed

Suff99 commented on common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/PhoneBoothDoorModel.java:
```diff
@@ -77,7 +77,7 @@ public void setupAnim(Entity entity, float f, float g, float h, float i, float j
 	@Override
 	public void setDoorPosition(boolean open) {
-		this.Door2.yRot = (open) ? -1.75f : 0;
+		this.Door2.yRot = (open) ? 1.75f : 0;
```
Fixed

Jeryn99 committed:
Some code review changes

Suff99 commented:
- [ ] BUG: Leaving the game and rejoining whilst inside the cave makes the growth door no longer function.
- [ ] BUG: Placing a door inside the root cave creates a portal for the growth shell
- [ ] BUG: Default shell's portal is not aligned with the interior view - doors are distinctively lower inside when looking from the outside
- [ ] BUG: You cannot view the external shell's model whilst inside - doors that open outwards are not visible.
- [ ] BUG: Door sounds can be heard twice when the door is closed
- [ ] BUG: Throwing items through the door will cause them to spaz out
- [ ] BUG: Breaking the shell in creative causes the model to no longer render until the next interaction.

Jeryn99 committed:
Re-instated Nuka

Jeryn99 committed:
Began figuring out why root door is being a pain

Jeryn99 committed:
Minor changes

Jeryn99 committed:
Merge remote-tracking branch 'origin/master' into feature/ip

Jeryn99 committed:
Conflict resolved

Jeryn99 committed:
- Added Cave Check
- Fixed Portals generating for root shell
- Fixed Terraformer working in Tardis's that are not caves
- Fixed Purple Crystal Console Pattern using old check logic
- Cleaned code in a few places where seen

Jeryn99 committed:
- Updated Immersive Portals

Jeryn99 committed:
BRIEFCASE

thejudge156 committed:
Various fixes, fix double-sound (#91)

Co-authored-by: BuildTools <unconfigured@null.spigotmc.org>

Jeryn99 committed:
Merge

50ap5ud5 commented on common/build.gradle:
```diff
@@ -5,6 +5,17 @@ dependencies {
     modCompileOnlyApi "net.minecraftforge:forgeconfigapiport-fabric:${rootProject.forge_config_api_port_version}"
+
+
+    modCompileOnlyApi('com.github.iPortalTeam.ImmersivePortalsMod:imm_ptl_core:v2.3.1-1.19') {
+        exclude(group: "net.fabricmc.fabric-api")
+        transitive(false)
+    }
+    modCompileOnlyApi('com.github.iPortalTeam.ImmersivePortalsMod:q_misc_util:v2.3.1-1.19') {
+        exclude(group: "net.fabricmc.fabric-api")
+        transitive(false)
+    }
+
```
Would be good to set a gradle variable for the IP dependencies

50ap5ud5 commented on common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisInteriorManager.java:
```diff
@@ -89,21 +105,19 @@ public CompoundTag saveData(CompoundTag tag) {
     }
 
     public void loadData(CompoundTag tag) {
+        System.out.println(tag);
```
Remove prints

50ap5ud5 commented on fabric/src/main/java/whocraft/tardis_refined/fabric/events/ModEvents.java:
```diff
@@ -38,12 +56,43 @@ public static void addCommonEvents() {
+        return uuids;
+    }
+
+    public static JsonElement getResponse(URL url) throws IOException {
```
What is this used for? Seems to not be called anywhere.

50ap5ud5 reviewed (changes requested):
Amazing work!

Jeryn99 committed:
Adjusted to 50s Feedback

Jeryn99 committed:
Immersive portals notice

Jeryn99 committed:
Fixed broken dims on old worlds

Jeryn99 committed:
Fixed buildscript change

Noklle commented:
Looks stunning, but I came across a weird issue

BUG: The collision of the door block prevents entry and exit. This can be circumvented in the overworld by jumping on top of the block, upon which you fall into it and can walk into the TARDIS, but this doesn't work when trying to exit the TARDIS. Works fine in spectator mode, given the lack of collision.

Suff99 commented:
It's been fixed pending release
